resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala

```diff
@@ -292,6 +292,14 @@ private[spark] object Config extends Logging {
       .checkValue(value => value > 0, "Allocation batch size should be a positive integer")
       .createWithDefault(5)
 
+  val KUBERNETES_MAX_PENDING_PODS =
+    ConfigBuilder("spark.kubernetes.allocation.max.pendingPods")
+      .doc("Maximum number of pending pods allowed during executor alloction for this application.")
+      .version("3.2.0")
+      .intConf
+      .checkValue(value => value > 0, "Maximum number of pending pods should be a positive integer")
+      .createWithDefault(150)
```

Comment:
This default seems high, can you explain why 150?

Reply:
Thanks for the review!

My main intention was to come up with a limit which protect us from overloading the k8s scheduler but still allows progressive upscaling. I think when a PODs spends long time in pending state we should make the allocation as early as possible but being careful to avoid the overloading of the k8s scheduler as that would be counterproductive for the allocations. And as we still use the batch size during upscaling there is a limited number of active new POD requests from a single Spark application (this also helps to avoid the overloading).

And the second reason was I hoped this a good default for those envs where the batchsize is already increased (I have seen examples where the batch size was set to 50). 

But I just run a few tests and although I have seen 150 pending PODs was not causing any problem during resource allocation my test was running in a EKS cluster where only one Spark app was submitted (my test app) and even the cluster size was small. 

So nevertheless we can go for a different solution:
**Another strategy to choose this limit would be to use a default which is conformed to the default batch size (which is really small = 5). So what about setting the default to 15 here?** In this case we can mention this new config in the migration guide. 

@holdenk WDYT?

Reply:
I'd like to propose to fallback to `spark.kubernetes.allocation.batch.size` by default. WDYT, @attilapiros and @holdenk ?

Reply:
@dongjoon-hyun 
> And, it would be great if we keep the existing test cases with the default configuration (disabled), and add new test coverage for this new conf.

So this PR has 3 small features which relates to each others:
- modify how the batch size limit is taken into account: earlier the next batch is not started when even one POD was stuck as newly created POD: **these causes some of the test changes**
- change what is outstanding PODs are: earlier Pending PODs and newly created PODs was counted as outstanding PODs which [was stopping the allocation if it was coming from the scheduler](https://github.com/apache/spark/blob/master/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala#L125). 
**This causes the rest of the unit test difference.**
- introduce limit for pending pods

Let me separate them into different PRs (at least for two) this will make the review easier.

Reply:
Sounds good, ping me on your split PRs and I'll be happy to take a look.